bdk_rpc/src/lib.rs

```diff
@@ -0,0 +1,377 @@
+        "main" => Network::Bitcoin,
+        "test" => Network::Testnet,
+        "regtest" => Network::Regtest,
+        "signet" => Network::Signet,
+        _ => {
+            return Err(RpcError::Generic(
+                "Invalid network in blockchain info".to_string(),
+            ))
+        }
+    };
+    if network != config.network {
+        return Err(RpcError::InvalidNetwork {
+            requested: config.network,
+            found: network,
+        });
+    }
+
+    Ok(client)
+}
+
+/// Sync the Tracker, Sparsechain and Client with each other.
+pub fn rpc_sync<K: Debug + Clone + Ord>(
+    client: &Client,
+    tracker: &mut KeychainTracker<K>,
+    sparse_chain: &mut SparseChain,
+) -> Result<(), RpcError> {
+    let start = std::time::Instant::now();
+
+    // Import spks from Tracker to Client
+    // Can have spks that are already imported
+    let spk_map = &*tracker.script_pubkeys();
+    let _labelled_scripts: Vec<(Option<String>, Script)> = spk_map
+        .iter()
+        .map(|((keychain, _), script)| {
+            let label = Some(format!("{:?}", keychain));
+            (label, script.clone())
+        })
+        .collect();
+
+    let labelled_scripts = _labelled_scripts
+        .iter()
+        .map(|(label, script)| {
+            let label = label.as_ref().map(|l| l.as_str());
+            (label, script)
+        })
+        .collect();
+    import_multi(&client, labelled_scripts)?;
+
+    // Import spks from Client back to Tracker
+    // This is useful when a fresh Tracker is started but it previously
+    // imported spks into Client.
+    let mut keychain_derive_map = Vec::new();
+    for (keychain, _) in &mut tracker.iter_keychains(..) {
+        let label = Value::String(format!("{:?}", keychain));
+        if let Ok(addresses) = client.call::<Value>("getaddressesbylabel", &[label]) {
+            let max_index = addresses
+                .as_object()
+                .expect("Address object expected")
+                .iter()
+                .count()
+                - 1;
+            keychain_derive_map.push((keychain, max_index));
+        }
+    }
+
+    for (keychain, max_index) in keychain_derive_map {
+        if tracker.derive_spks(keychain, max_index as u32) == true {
+            debug!("Added new sks from core wallet, upto index : {}", max_index);
+        } else {
+            debug!("Core wallet is in sync with tracker");
+        }
+    }
+
+    // Check for reorg. Invalidate checkpoints from reorg height.
+    let mut last_common = None;
+    let best_hash = client.get_best_block_hash()?;
+    for checkpoint in sparse_chain.iter_checkpoints(..).rev() {
+        if checkpoint.hash == client.get_block_hash(checkpoint.height as u64)?
+            && checkpoint.hash != best_hash
+        {
+            last_common = Some(checkpoint);
+        }
+    }
+
+    if let Some(checkpoint) = last_common {
+        sparse_chain.invalidate_checkpoint(checkpoint.height);
```

Comment:
why are you invalidating the last common one? If they have it in common it is not invalid.

Also invalidate checkpoint is meant to be a private API.